File: ng2-ts/src/app/service/kudos.service.ts
import { Injectable }    from '@angular/core';
import { Headers, Http } from '@angular/http';

import { Observable } from 'rxjs';

import { Kudos } from '../model/kudos';

@Injectable()
export class KudosService {
    
    private restUrl = '/rest/kudos/';
    
    constructor(private http: Http) {}
    
    getAllKudos() {
        return this.http.get(this.restUrl)
            .map( (response:any) => {
                if (response.status == 200) {
                    return response.json() as Kudos[];
                } else {
                    return null
                }
            });
    }
    
    getKudos(id: number) {
        let url = this.restUrl + id;
        return this.http.get(url)
        	.map( (response:any) => {
        		if (response.status == 200) {
        			return response.json() as Kudos;
        		} else {
        			return null
        		}
        	});
    }
    
    save(kudos: Kudos) {
        if (kudos.id) {
            return this.put(kudos);
        }
        return this.post(kudos);
    }
    
    delete(kudos: Kudos) {
        let headers = new Headers();
        headers.append('Content-Type', 'application/json');
        let url = this.restUrl + kudos.id;
        return this.http
            .delete(url, {
                headers: headers
            })
            .map( (response:any) => {
            	if (response.status == 200) {
            		return true;
            	} else {
            		return false;
            	}
            });
    }

    private post(kudos: Kudos) {
        let headers = new Headers({
            'Content-Type': 'application/json'
        });
        return this.http
            .post(this.restUrl, JSON.stringify(kudos), {
                headers: headers
            })
            .map( (response:any) => {
            	if (response.status == 200) {
            		return response.json() as Kudos;
            	}
            });
    }

    private put(kudos: Kudos) {
        let headers = new Headers();
        headers.append('Content-Type', 'application/json');
        let url = this.restUrl;
        return this.http
            .put(url, JSON.stringify(kudos), {
                headers: headers
            })
            .map( (response:any) => {
            	if (response.status == 200) {
            		return response.json() as Kudos;
            	} else {
            		return null;
            	}
            });
    }
}

File: ng2-dart/lib/service/kudos.service.dart
import 'dart:async';
import 'dart:convert';

import 'package:angular2/core.dart';
import 'package:http/http.dart';

import '../model/kudos.dart';

@Injectable()
class KudosService {

	static final _headers = { 'Content-Type' : 'application/json' };
	static const _restUrl = '/rest/kudos/';

	final Client _http;

	KudosService(this._http);

	Future<List<Kudos>> getAllKudos() async {
		try {
			final response = await _http.get(_restUrl);
			final kudoss = _extractData(response)
			.map((value) => new Kudos.fromJson(value))
			.toList();

			return kudoss;
		} catch (e) {
			throw _handleError(e);
		}
	}

	Future<Kudos> getKudos(int id) async {
		var url = '$_restUrl${id}';
		try {
			final response = await _http.get(url);
			final kudos = _extractData(response)
			.map((value) => new Kudos.fromJson(value));

			return kudos;
		} catch (e) {
			throw _handleError(e);
		}
	}

	Future<Kudos> save(Kudos kudos) async {
		if (kudos.id != null) {
			return this._put(kudos);
		} else {
			return this._post(kudos);
		}
	}	

	Future<Kudos> delete(int id) async {
		var url = '$_restUrl${id}';
		try {
			final response = await _http.delete(url);
			final kudos = _extractData(response)
			.map((value) => new Kudos.fromJson(value));

			return kudos;
		} catch (e) {
			throw _handleError(e);
		}
	}

	Future<Kudos> _post(Kudos kudos) async {
		try {
     		final response = await _http.post(_restUrl,
         	headers: _headers, body: JSON.encode(kudos));
      
      		return new Kudos.fromJson(_extractData(response));
   		} catch (e) {
      		throw _handleError(e);
    	}
	}

	Future<Kudos> _put(Kudos kudos) async {
		try {
     		final response = await _http.put(_restUrl,
         	headers: _headers, body: JSON.encode(kudos));
      
      		return new Kudos.fromJson(_extractData(response));
   		} catch (e) {
      		throw _handleError(e);
    	}
	}

	dynamic _extractData(Response resp) => JSON.decode(resp.body)['data'];

	Exception _handleError(dynamic e) {
    	print(e); // for demo purposes only
    	return new Exception('Server error; cause: $e');
  	}

}